File: Grind75_1/week2/FirstBadVersion.py
"""
https://leetcode.com/problems/first-bad-version/description/
"""

# The isBadVersion API is already defined for you.
# def isBadVersion(version: int) -> bool:

class Solution:

    def firstBadVersionRec(self, left: int, right: int) -> int:
        # print (str(left) + ' : ' + str(right))
        # base case
        if (left >= right):
            return left

        mid = left + floor((right - left) / 2)

        if (isBadVersion(mid)):
            return self.firstBadVersionRec(left, mid)
        else:
            return self.firstBadVersionRec(mid + 1, right)


    def firstBadVersionStd(self, n: int) -> int:
        a = 1
        # use b to keep track of earliest bad version found so far
        b = n

        # searching for left-most bad version
        while (a < b):
            mid = floor(a + ((b - a) / 2))
            if (isBadVersion(mid)):
                b = mid
            else:
                # mid is a good version, start from one to the right
                a = mid + 1
        
        return a

    def firstBadVersion(self, n: int) -> int:
        # return self.firstBadVersionStd(n)
        return self.firstBadVersionRec(1, n)
            
            
        

File: Grind75/Binary-Search/first-bad-version.js
// https://leetcode.com/problems/first-bad-version/description/

/**
 * Definition for isBadVersion()
 * 
 * @param {integer} version number
 * @return {boolean} whether the version is bad
 * isBadVersion = function(version) {
 *     ...
 * };
 */

/**
Binary search and maintain var that tracks the most left bad version
if bad version found
    earliestBad = mid
    go left in search for a bad version even earlier
else
    go right since mid and earlier are all good versions

- Time: O(log n)
- Space: O(1)
 */

/**
 * @param {function} isBadVersion()
 * @return {function}
 */
var solution = function(isBadVersion) {
    /**
     * @param {integer} n Total versions
     * @return {integer} The first bad version
     */
    return function(n) {
        let earliestBad = n
        let l = 0
        let r = n - 1
        while (l <= r) {
            const mid = Math.floor((r - l) / 2) + l

            if (isBadVersion(mid) === true) {
                earliestBad = mid
                r = mid - 1
            } else {
                l = mid + 1
            }
        }

        return earliestBad
    };
};